pennylane/ops/functions/bind_new_parameters.py

```diff
@@ -0,0 +1,149 @@
+
+    if invalid_params:
+        raise ValueError(
+            "The length of the new parameters does not match the expected shape; "
+            f"got {len(params)}, expected {op.num_params}."
+        )
+
+    return params
+
+
+@singledispatch
+def bind_new_parameters(op: Operator, params: Optional[Sequence[TensorLike]] = None) -> Operator:
+    """Create a new operator with updated parameters
+
+    This function takes an :class:`~.Operator` and new parameters as input and
+    returns a new operator of the same type with the new parameters. This function
+    does not mutate the original operator.
+
+    Args:
+        op (.Operator): Operator to update
+        params (Sequence[TensorLike]): New parameters to create operator with
+
+    Returns:
+        .Operator: New operator with updated parameters
+
+    Raises:
+        ValueError: If the shape of the old and new operator parameters don't match
+    """
+    params = _validate_params(op, params)
+
+    return op.__class__(*params, wires=op.wires, **copy.deepcopy(op.hyperparameters))
+
+
+# pylint: disable=missing-docstring
+@bind_new_parameters.register
+def bind_new_parameters_composite_op(
+    op: CompositeOp, params: Optional[Sequence[TensorLike]] = None
+):
+    params = _validate_params(op, params)
+    new_operands = []
+
+    for operand in op.operands:
+        sub_params = params[: operand.num_params]
+        params = params[operand.num_params :]
+        new_operands.append(bind_new_parameters(operand, sub_params))
+
+    return op.__class__(*new_operands, **copy.deepcopy(op.hyperparameters))
+
+
+# pylint: disable=missing-docstring
+@bind_new_parameters.register
+def bind_new_parameters_symbolic_op(op: SymbolicOp, params: Optional[Sequence[TensorLike]] = None):
+    params = _validate_params(op, params)
+    new_scalar = []
+
+    if isinstance(op, ScalarSymbolicOp):
+        new_scalar.append(params[0])
+        params = params[1:]
+
+    new_base = bind_new_parameters(op.base, params)
+    new_hyperparameters = copy.deepcopy(op.hyperparameters)
+    _ = new_hyperparameters.pop("base")
+
+    if isinstance(op, Adjoint):
+        # Need this branch because using the other class signature results in a call to
+        # `Adjoint.__new__` which doesn't raise an error but does return an unusable
+        # object
+        return Adjoint(new_base)
+    try:
+        # `try-except` block to accomodate for the fact that different `ScalarSymbolicOp`
+        # subclasses accept the base and scalar in different orders
+        return op.__class__(new_base, *new_scalar, **new_hyperparameters)
+    except AttributeError:
+        pass
+
+    return op.__class__(*new_scalar, new_base, **new_hyperparameters)
+
+
+# pylint: disable=missing-docstring
+@bind_new_parameters.register
+def bind_new_parameters_hamiltonian(
+    op: qml.Hamiltonian, params: Optional[Sequence[TensorLike]] = None
+):
+    params = _validate_params(op, params)
+    new_observables = copy.deepcopy(op.ops)
```

Comment:
So copies and deep copies are useful when there's a chance the original could be mutated by something else and we don't want to be affected by that later mutation.

If we move to the system where we never in-place mutate anything, we can trust that the `op.ops` will look the exact same for the rest of the program.  We don't have to worry about creating a new `Hamiltonian` with those same objects, and then the objects changing on us.

Therefore, we don't need to perform `deepcopy` here (or above either).